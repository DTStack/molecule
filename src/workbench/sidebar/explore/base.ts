import {
    getBEMElement,
    getBEMModifier,
    prefixClaName,
} from 'mo/common/className';
import {
    ID_ACTIVITY_BAR,
    ID_SIDE_BAR,
    ID_EXPLORER,
    ID_FOLDER_TREE,
    ID_EDITOR_TREE,
} from 'mo/common/id';

const defaultClassName = prefixClaName(ID_SIDE_BAR);
const defaultExplorerClassName = prefixClaName(ID_EXPLORER, defaultClassName);
const activityBarItemFloatClassName = getBEMModifier(
    getBEMElement(defaultExplorerClassName, ID_ACTIVITY_BAR),
    'float'
);

const folderTreeClassName = prefixClaName(ID_FOLDER_TREE);
const folderTreeInputClassName = getBEMModifier(folderTreeClassName, 'input');
const folderTreeEditClassName = getBEMModifier(folderTreeClassName, 'editable');

const editorTreeClassName = prefixClaName(ID_EDITOR_TREE);
const editorTreeItemClassName = getBEMElement(editorTreeClassName, 'item');
const editorTreeGroupClassName = getBEMElement(editorTreeClassName, 'group');
const editorTreeFileNameClassName = getBEMElement(
    editorTreeItemClassName,
    'fileName'
);
const editorTreeFilePathClassName = getBEMElement(
    editorTreeItemClassName,
    'filePath'
);
const editorTreeActiveItemClassName = getBEMModifier(
    editorTreeItemClassName,
    'active'
);
const editorTreeCloseIconClassName = getBEMElement(
    editorTreeClassName,
    'close'
);
const editorTreeFileIconClassName = getBEMElement(editorTreeClassName, 'file');

export {
    defaultExplorerClassName,
    activityBarItemFloatClassName,
    folderTreeClassName,
    folderTreeInputClassName,
    folderTreeEditClassName,
    editorTreeClassName,
    editorTreeItemClassName,
    editorTreeGroupClassName,
    editorTreeFileNameClassName,
    editorTreeFilePathClassName,
    editorTreeActiveItemClassName,
    editorTreeCloseIconClassName,
    editorTreeFileIconClassName,
};
