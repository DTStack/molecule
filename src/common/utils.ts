export function searchById(id) {
    return (item) => item.id === id;
}

/**
 * Clone a new object by an object instance
 * @param origin Original object instance
 */
export function cloneInstance<T>(origin: T) {
    try {
        const prototypes = Object.getPrototypeOf(origin) || {};
        Object.keys(prototypes).forEach((prop) => {
            if (typeof prototypes[prop] === 'function') {
                prototypes[prop].bind(origin);
            }
        });
        return {
            ...origin,
            ...prototypes,
        };
    } catch (e) {
        console.error('Function cloneInstance error:', e);
    }
}

/**
 * Merge multiple functions to one function
 * @param funcs
 */
export function mergeFunctions(...funcs) {
    return function (...args) {
        funcs.filter((fn) => !!fn).forEach((fn) => fn?.(...args));
    };
}

export function randomId() {
    return Date.now() + Math.round(Math.random() * 1000);
}

export function mergeObjects<T>(source: object, target: object): T {
    for (const prop in target) {
        if (typeof source[prop] === 'object' && target[prop]) {
            source[prop] = mergeObjects(source[prop], target[prop]);
        } else {
            source[prop] = target[prop];
        }
    }
    return source as any;
}

/**
 * It's converts an object to a flatted object,
 * eg: { a: { b: 'test' }}, result is : { 'a.b': 'test' }
 * @param target flat target
 */
export function flatObject(target: object): object {
    const flatted = {};
    const recurse = (parent: string, target: object) => {
        for (const prop in target) {
            if (prop) {
                const value = target[prop];
                const path = parent ? `${parent}.${prop}` : prop;
                if (typeof value === 'object') {
                    recurse(path, value);
                } else {
                    flatted[path] = value;
                }
            }
        }
    };
    recurse('', target);
    return flatted;
}

/**
 * It's converts a flatted object to a normal object,
 *  eg: { 'a.b': 'test' }, result is : { a: { b: 'test' }}
 * @param target flat target
 */
export function normalizeFlattedObject(target: object): object {
    let normalized = {};

    for (const prop in target) {
        if (prop) {
            const flattedProps = prop.split('.');
            let prevProp = '';
            let root = {};
            let prevObj = root;
            for (let i = 0; i < flattedProps.length; ++i) {
                if (i === flattedProps.length) break;
                const key = flattedProps[i];
                const current = { [key]: '' };

                if (prevProp) {
                    prevObj[prevProp] = current;
                }
                prevObj = current;
                prevProp = key;
                if (i === 0) {
                    root = current;
                }
            }
            prevObj[prevProp] = target[prop];
            normalized = mergeObjects(normalized, root);
        }
    }
    return normalized;
}

/**
 * Determine if a color is light or dark.
 * @param color HEX or RGB
 */
export function colorLightOrDark(color: string) {
    // Variables for red, green, blue values
    let r: number;
    let g: number;
    let b: number;

    // Check the format of the color, HEX or RGB?
    if (color.match(/^rgb/)) {
        // If RGB --> store the red, green, blue values in separate variables
        const matchArray =
            color.match(
                /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
            ) || [];
        r = +matchArray[1];
        g = +matchArray[2];
        b = +matchArray[3];
    } else {
        // If hex --> Convert it to RGB
        let rgbNum = +('0x' + color.slice(1, 7));
        if (color.length < 5) {
            rgbNum = +('0x' + color.slice(1).replace(/./g, '$&$&').slice(0, 6));
        }
        r = rgbNum >> 16;
        g = (rgbNum >> 8) & 255;
        b = rgbNum & 255;
    }

    // HSP (Highly Sensitive Poo) equation from http://alienryderflex.com/hsp.html
    const hsp = Math.sqrt(0.299 * (r * r) + 0.587 * (g * g) + 0.114 * (b * b));

    // Using the HSP value, determine whether the color is light or dark
    if (hsp > 127.5) {
        return 'light';
    }
    return 'dark';
}
