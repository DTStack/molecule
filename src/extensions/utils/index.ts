import type { IMenuItemProps, IterableItem } from 'mo/types';

export function createMenuDuplicate(item: IMenuItemProps): IMenuItemProps {
    return {
        ...item,
        id: `${item.id}_duplicate`,
        name: `隐藏 “${item.name}”`,
        icon: undefined,
        symbolic: item.id,
    };
}

export function createContextMenu(items: IterableItem[]): IMenuItemProps[] {
    return items.map((item) => ({
        id: item.id,
        name: item.name,
        disabled: item.disabled,
        icon: item.hidden ? undefined : 'check',
    }));
}

export function isInputEle() {
    return document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
}

/**
 * Judge current active element is monaco-editor's textarea or native element
 */
export function isNativeWidget() {
    const isEditableEle =
        document.activeElement instanceof HTMLInputElement || document.activeElement instanceof HTMLTextAreaElement;
    if (!isEditableEle) return true;
    return !document.activeElement?.classList.contains('inputarea');
}
