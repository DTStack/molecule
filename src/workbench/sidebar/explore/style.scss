@import 'mo/style/common';

#{$sidebar} {
    #{$collapse}__header {
        &__title {
            font-weight: 700;
            text-transform: uppercase;
        }
    }

    #{$folderTree} {
        #{$tree}__indent__guide {
            opacity: 0;
            transition: opacity 0.3s linear;
        }

        &:hover #{$tree}__indent__guide {
            opacity: 1;
        }
    }
}

#{$folderTree} {
    min-height: 100%;
    padding-bottom: 22px;

    &--editable {
        background: rgba(0, 0, 0, 0.3);

        #{$folderTree}--input {
            background: var(--input-background);
            border: 1px solid var(--list-focusOutline);
            color: var(--input-foreground);
            font-family: inherit;
            font-size: 13px;
            height: 20px;
            left: 0;
            margin: 0;
            outline: none;
            padding: 0;
            position: relative;
            width: calc(100% - 2px);

            &:focus,
            &:active {
                border-color: var(--list-focusOutline);
                outline: inherit;
            }

            &::selection {
                background: rgb(90, 126, 167);
            }

            &::placeholder {
                color: var(--input-placeholderForeground);
            }
        }
    }
}

#{$editorTree} {
    font-size: 13px;

    &__item {
        // keep same with tree node
        align-items: center;
        border: 1px solid transparent;
        cursor: pointer;
        display: flex;
        height: 22px;
        line-height: 22px;
        list-style: none;
        margin: 0;
        outline: 0;
        padding-left: 16px;
        user-select: none;
        white-space: nowrap;

        &--active {
            // 这里的 inactive 是相对于 focus 的 active 来说的，是相对的 inactive
            background-color: var(--list-inactiveSelectionBackground);

            #{$editorTree}__close {
                color: var(--icon-foreground);
                opacity: 1;
            }
        }

        &:hover {
            &:not(#{$editorTree}__item--active) {
                background-color: var(--list-hoverBackground);
            }

            #{$editorTree}__close {
                opacity: 1;
            }
        }

        &:focus {
            background: var(--list-activeSelectionBackground);
            border-color: var(--list-focusOutline);
            color: #fff;
        }

        &__fileName {
            flex-basis: auto;
            margin-left: 5px;
            overflow: hidden;
            text-overflow: ellipsis;
            white-space: pre;
        }

        &__filePath {
            flex: 1;
            font-size: 0.9em;
            margin-left: 5px;
            opacity: 0.8;
            overflow: hidden;
            padding-top: 2px;
            text-overflow: ellipsis;
            white-space: pre;
        }
    }

    &__group {
        align-items: center;
        display: flex;
        font-size: 11px;
        height: 22px;
        line-height: 22px;
        padding-left: 16px;
        user-select: none;

        &:hover {
            background-color: var(--list-hoverBackground);

            #{$toolBar} {
                opacity: 1;
            }
        }

        #{$toolBar} {
            margin-left: auto;
            opacity: 0;

            .codicon {
                color: var(--activityBar-inactiveForeground);

                &:hover {
                    color: var(--activityBar-activeBorder);
                }
            }
        }
    }

    &__close {
        color: var(--icon-foreground);
        margin-right: 2px;
        opacity: 0;

        &:hover {
            opacity: 1;
        }
    }

    &__file[class*='codicon-'] {
        flex-shrink: 0;
        height: 22px;
        line-height: 23px;
        width: 16px;
    }
}
