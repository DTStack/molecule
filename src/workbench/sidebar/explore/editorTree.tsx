import React, { useLayoutEffect, useRef } from 'react';
import { IEditorTreeController } from 'mo/controller';
import { IEditor, IEditorGroup } from 'mo/model';
import {
    IActionBarItemProps,
    Icon,
    IMenuItemProps,
    ITabProps,
    Menu,
    Scrollable,
    Toolbar,
    useContextView,
} from 'mo/components';
import {
    editorTreeActiveItemClassName,
    editorTreeClassName,
    editorTreeCloseIconClassName,
    editorTreeFileIconClassName,
    editorTreeFileNameClassName,
    editorTreeFilePathClassName,
    editorTreeGroupClassName,
    editorTreeItemClassName,
} from './base';
import { classNames } from 'mo/common/className';
import { getEventPosition } from 'mo/common/dom';
import { localize } from 'mo/i18n/localize';
import {
    ICollapseItem,
    HEADER_HEIGTH,
    MAX_GROW_HEIGHT,
} from 'mo/components/collapse';
import Scrollbar from 'react-scrollbars-custom';
import { constants } from 'mo/services/builtinService/const';
import type { UniqueId } from 'mo/common/types';

// override onContextMenu
type UnionEditor = Omit<
    IEditor & IEditorTreeController,
    'onContextMenu' | 'initView'
>;
export interface IOpenEditProps extends UnionEditor {
    /**
     * Group Header toolbar
     */
    groupToolbar?: IActionBarItemProps<IEditorGroup>[];
    /**
     * Item context menus
     */
    contextMenu?: IMenuItemProps[];
    /**
     * Group Header context menus
     * It'll use the value of contextMenu if specify contextMenu but not specify headerContextMenu
     */
    headerContextMenu?: IMenuItemProps[];
    onContextMenu?: (
        menu: IMenuItemProps,
        groupId: UniqueId,
        file?: ITabProps
    ) => void;
    panel: ICollapseItem;
}

const EditorTree = (props: IOpenEditProps) => {
    const {
        current,
        groups,
        groupToolbar,
        contextMenu = [],
        headerContextMenu,
        panel,
        onSelect,
        onSaveGroup,
        onContextMenu,
        onCloseGroup,
        onClose,
        onToolbarClick,
    } = props;

    const wrapper = useRef<HTMLDivElement>(null);
    const scrollable = useRef<Scrollbar>(null);

    // scroll into view
    useLayoutEffect(() => {
        const scrollHeight = scrollable.current?.scrollHeight || 0;
        if (scrollHeight > MAX_GROW_HEIGHT - HEADER_HEIGTH) {
            const activeItem = wrapper.current?.querySelector<HTMLDivElement>(
                `.${editorTreeActiveItemClassName}`
            );
            if (activeItem) {
                const top = activeItem.offsetTop;
                scrollable.current?.scrollTo(0, top);
            }
        }
    }, [current?.id && current.tab?.id]);

    if (!groups || !groups.length) return null;

    const contextView = useContextView();

    const handleCloseClick = (group: IEditorGroup, file: ITabProps) => {
        onClose?.(file.id!, group.id!);
    };

    const handleItemClick = (group: IEditorGroup, file: ITabProps) => {
        if (group.id !== current?.id || file.id !== current?.tab?.id) {
            onSelect?.(file.id!, group.id!);
        }
    };

    const handleOnMenuClick = (
        menu: IMenuItemProps,
        group: IEditorGroup,
        file?: ITabProps
    ) => {
        contextView.hide();
        onContextMenu?.(menu, group.id!, file);
    };

    const handleRightClick = (
        e: React.MouseEvent<HTMLDivElement, MouseEvent>,
        group: IEditorGroup,
        file: ITabProps
    ) => {
        e.preventDefault();
        contextView.show(getEventPosition(e), () => (
            <Menu
                role="menu"
                onClick={(_, item) => handleOnMenuClick(item!, group, file)}
                data={contextMenu}
            />
        ));
    };

    const handleHeaderRightClick = (
        e: React.MouseEvent<HTMLDivElement, MouseEvent>,
        group: IEditorGroup
    ) => {
        e.preventDefault();
        const groupHeaderContext = headerContextMenu || contextMenu;
        contextView.show(getEventPosition(e), () => (
            <Menu
                role="menu"
                onClick={(_, item) => handleOnMenuClick(item!, group)}
                data={groupHeaderContext}
            />
        ));
    };

    // click group title will open the first file in this group
    const handleGroupClick = (e, group: IEditorGroup) => {
        const { target } = e;
        const firstFile = group.data?.[0];
        if (target.nextElementSibling && firstFile) {
            onSelect?.(firstFile.id!, group.id!);
            target.nextElementSibling.focus();
        }
    };

    const handleToolBarClick = (
        e: React.MouseEvent<Element, MouseEvent>,
        item: IActionBarItemProps,
        group: IEditorGroup
    ) => {
        e.stopPropagation();
        switch (item.id) {
            case constants.EXPLORER_TOGGLE_CLOSE_GROUP_EDITORS:
                onCloseGroup?.(group.id!);
                break;
            case constants.EXPLORER_TOGGLE_SAVE_GROUP:
                onSaveGroup?.(group.id!);
                break;
            default:
                // default behavior
                onToolbarClick?.(item, group.id!);
                break;
        }
    };

    return (
        <Scrollable noScrollX isShowShadow ref={scrollable}>
            <div
                className={editorTreeClassName}
                ref={wrapper}
                data-content={panel.id}
            >
                {groups.map((group, index) => {
                    return (
                        <React.Fragment key={index}>
                            {groups.length !== 1 && (
                                <div
                                    className={editorTreeGroupClassName}
                                    onClick={(e) => handleGroupClick(e, group)}
                                    onContextMenu={(e) =>
                                        handleHeaderRightClick(e, group)
                                    }
                                    key={index}
                                >
                                    {localize(
                                        'sidebar.explore.openEditor.group',
                                        'Group',
                                        (index + 1).toString()
                                    )}
                                    {groupToolbar && (
                                        <Toolbar
                                            data={groupToolbar}
                                            onClick={(e, item) =>
                                                handleToolBarClick(
                                                    e,
                                                    item,
                                                    group
                                                )
                                            }
                                        />
                                    )}
                                </div>
                            )}
                            {group.data?.map((file) => {
                                const isActive =
                                    group.id === current?.id &&
                                    file.id === current?.tab?.id;
                                return (
                                    <div
                                        title={
                                            file.data?.path &&
                                            `${file.data?.path}/${file.name}`
                                        }
                                        className={classNames(
                                            editorTreeItemClassName,
                                            isActive &&
                                                editorTreeActiveItemClassName
                                        )}
                                        tabIndex={0}
                                        key={`${index}_${file.id}`}
                                        onClick={() =>
                                            handleItemClick(group, file)
                                        }
                                        onContextMenu={(e) =>
                                            handleRightClick(e, group, file)
                                        }
                                    >
                                        <Icon
                                            className={
                                                editorTreeCloseIconClassName
                                            }
                                            onClick={() =>
                                                handleCloseClick(group, file)
                                            }
                                            type="close"
                                        />
                                        <Icon
                                            className={
                                                editorTreeFileIconClassName
                                            }
                                            type={file.data?.icon || file.icon}
                                        />
                                        <span
                                            className={
                                                editorTreeFileNameClassName
                                            }
                                        >
                                            {file.name}
                                        </span>
                                        <span
                                            className={
                                                editorTreeFilePathClassName
                                            }
                                        >
                                            {file.data?.path}
                                        </span>
                                    </div>
                                );
                            })}
                        </React.Fragment>
                    );
                })}
            </div>
        </Scrollable>
    );
};

export { EditorTree };
