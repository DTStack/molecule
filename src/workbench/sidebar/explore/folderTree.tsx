import 'reflect-metadata';
import React, { memo, useRef, useEffect, useLayoutEffect } from 'react';
import { IFolderTree, IFolderTreeNodeProps } from 'mo/model';
import { select, getEventPosition } from 'mo/common/dom';
import Tree from 'mo/components/tree';
import { IMenuItemProps, Menu } from 'mo/components/menu';
import { Button } from 'mo/components/button';
import type { IFolderTreeController } from 'mo/controller/explorer/folderTree';
import { useContextView } from 'mo/components/contextView';
import { useContextMenu } from 'mo/components/contextMenu';
import {
    folderTreeClassName,
    folderTreeEditClassName,
    folderTreeInputClassName,
} from './base';
import { classNames } from 'mo/common/className';
import { Scrollable } from 'mo/components';
import { ICollapseItem } from 'mo/components/collapse';

export interface IFolderTreeProps extends IFolderTreeController, IFolderTree {
    panel: ICollapseItem;
}

const detectHasEditableStatus = (data) => {
    const stack = [...data];
    let res = false;
    while (stack.length) {
        const headElm = stack.pop();
        if (headElm?.isEditable) {
            res = true;
            break;
        } else {
            stack.push(...(headElm?.children || []));
        }
    }
    return res;
};

/**
 * A simple wrapper Input, achieve autoFucus & auto select file name
 */
const Input = React.forwardRef(
    (
        // same as raw input
        props: React.DetailedHTMLProps<
            React.InputHTMLAttributes<HTMLInputElement>,
            HTMLInputElement
        >,
        ref: React.LegacyRef<HTMLInputElement>
    ) => {
        const inputRef = useRef<HTMLInputElement>(null);
        useLayoutEffect(() => {
            if (inputRef.current) {
                const ext = ((props.defaultValue || '') as string).lastIndexOf(
                    '.'
                );
                inputRef.current.selectionStart = 0;
                inputRef.current.selectionEnd =
                    // if period at position of 0, then this period means hidden file
                    ext > 0
                        ? ext
                        : ((props.defaultValue || '') as string).length;
            }
        }, []);
        return <input {...props} ref={inputRef} />;
    }
);

const FolderTree: React.FunctionComponent<IFolderTreeProps> = (props) => {
    const {
        folderTree = {},
        entry,
        panel,
        onUpdateFileName,
        onSelectFile,
        onDropTree,
        onClickContextMenu,
        onRightClick,
        onLoadData,
        createTreeNode,
        onExpandKeys,
        ...restProps
    } = props;

    const {
        data = [],
        folderPanelContextMenu = [],
        expandKeys,
        loadedKeys,
        current,
    } = folderTree;

    const inputRef = useRef<HTMLInputElement>(null);
    // tree context view
    const contextMenu = useRef<ReturnType<typeof useContextMenu>>();

    // panel context view
    const contextView = useContextView();

    // to detect current tree whether is editable
    const hasEditable = detectHasEditableStatus(data);

    const onClickMenuItem = (e, item) => {
        onClickContextMenu?.(item);
        contextMenu.current?.hide();
    };

    // init context menu
    const initContextMenu = () => {
        return useContextMenu({
            anchor: select(`.${folderTreeClassName}`),
            render: () => (
                <Menu
                    role="menu"
                    onClick={onClickMenuItem}
                    data={folderPanelContextMenu}
                />
            ),
        });
    };

    const handleMenuClick = (
        item: IMenuItemProps,
        data: IFolderTreeNodeProps
    ) => {
        onClickContextMenu?.(item, data);
        contextView.hide();
    };

    const handleRightClick = (event, data) => {
        const menuItems = onRightClick?.(data) || [];

        menuItems.length &&
            contextView?.show(getEventPosition(event), () => (
                <Menu
                    role="menu"
                    onClick={(_, item) => handleMenuClick(item!, data)}
                    data={menuItems}
                />
            ));
    };

    const handleUpdateFile = (
        e: HTMLInputElement,
        node: IFolderTreeNodeProps
    ) => {
        const newName = e.value;
        onUpdateFileName?.({
            ...node,
            name: newName,
        });
    };

    /**
     * update file info when input blur
     */
    const handleInputBlur = (
        e: React.FocusEvent<HTMLInputElement>,
        node: IFolderTreeNodeProps
    ) => {
        handleUpdateFile(e.target, node);
    };

    /**
     * update file info when press `Enter` or `esc`
     */
    const handleInputKeyDown = (
        e: React.KeyboardEvent<HTMLInputElement>,
        node: IFolderTreeNodeProps
    ) => {
        if (e.keyCode === 13 || e.keyCode === 27) {
            handleUpdateFile(e.target as EventTarget & HTMLInputElement, node);
        }
    };

    const renderTitle = (node: IFolderTreeNodeProps) => {
        const { isEditable, name } = node;

        return isEditable ? (
            <Input
                role="input"
                className={folderTreeInputClassName}
                type="text"
                defaultValue={name}
                ref={inputRef}
                onKeyDown={(e) => handleInputKeyDown(e, node)}
                autoComplete="off"
                autoFocus
                onBlur={(e) => handleInputBlur(e, node)}
                onClick={(e) => e.stopPropagation()}
            />
        ) : (
            name!
        );
    };

    const handleTreeClick = () => {
        onSelectFile?.();
    };

    const handleDropTree = (source, target) => {
        onDropTree?.(source, target);
    };

    const handleAddRootFolder = () => {
        createTreeNode?.('RootFolder');
    };

    useEffect(() => {
        if (folderPanelContextMenu.length > 0) {
            contextMenu.current = initContextMenu();
        }
        return () => {
            contextMenu.current?.dispose();
        };
    }, [data.length]);

    const welcomePage = (
        <div data-content={panel.id}>
            {entry ? (
                <>{entry}</>
            ) : (
                <div style={{ padding: '10px 5px' }}>
                    you have not yet opened a folder
                    <Button onClick={handleAddRootFolder}>Add Folder</Button>
                </div>
            )}
        </div>
    );

    if (!data.length) return welcomePage;

    return (
        <Scrollable noScrollX isShowShadow>
            <div data-content={panel.id} style={{ height: '100%' }}>
                <Tree
                    // root folder do not render
                    activeKey={current?.id}
                    expandKeys={expandKeys}
                    loadedKeys={loadedKeys}
                    data={data[0]?.children || []}
                    className={classNames(
                        folderTreeClassName,
                        hasEditable && folderTreeEditClassName
                    )}
                    draggable={!hasEditable}
                    onDropTree={handleDropTree}
                    onSelect={onSelectFile}
                    onTreeClick={handleTreeClick}
                    onRightClick={handleRightClick}
                    renderTitle={renderTitle}
                    onLoadData={onLoadData}
                    onExpand={onExpandKeys}
                    {...restProps}
                />
            </div>
        </Scrollable>
    );
};
export default memo(FolderTree);
