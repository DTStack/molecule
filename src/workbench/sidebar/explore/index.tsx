import 'reflect-metadata';
import { connect } from 'mo/react';
import { container } from 'tsyringe';
import { FolderTreeService } from 'mo/services';
import { Explorer } from './explore';
import FolderTree from './folderTree';
import { FolderTreeController } from 'mo/controller/explorer/folderTree';
import { EditorTree } from './editorTree';

const folderTreeService = container.resolve(FolderTreeService);
const folderTreeController = container.resolve(FolderTreeController);

const FolderTreeView = connect(
    folderTreeService,
    FolderTree,
    folderTreeController
);

export { Explorer, FolderTreeView, FolderTree, EditorTree };
