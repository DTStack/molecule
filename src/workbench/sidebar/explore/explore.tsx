import React from 'react';
import { Collapse } from 'mo/components/collapse';
import { Header, Content } from 'mo/workbench/sidebar';
import { IExplorer } from 'mo/model/workbench/explorer/explorer';
import { IExplorerController } from 'mo/controller/explorer/explorer';
import { Toolbar } from 'mo/components/toolbar';
import { defaultExplorerClassName } from './base';
import { localize } from 'mo/i18n/localize';

type IExplorerProps = IExplorer & IExplorerController;

export const Explorer: React.FunctionComponent<IExplorerProps> = (
    props: IExplorerProps
) => {
    const {
        activePanelKeys,
        data = [],
        headerToolBar,
        onClick,
        onActionsContextMenuClick,
        onCollapseChange,
        onToolbarClick,
    } = props;
    return (
        <div className={defaultExplorerClassName}>
            <Header
                title={localize('sidebar.explore.title', 'Explorer')}
                toolbar={
                    <Toolbar
                        data={[headerToolBar!]}
                        onClick={onClick}
                        onContextMenuClick={onActionsContextMenuClick}
                    />
                }
            />
            <Content>
                <Collapse
                    data={data}
                    activePanelKeys={activePanelKeys}
                    onCollapseChange={onCollapseChange}
                    onToolbarClick={onToolbarClick}
                />
            </Content>
        </div>
    );
};
