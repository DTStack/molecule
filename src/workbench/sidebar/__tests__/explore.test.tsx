import React from 'react';
import { create, act } from 'react-test-renderer';
import { cleanup, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { expectFnCalled } from '@test/utils';
import { Explorer } from '../explore';

const mockContextMenu = {
    id: 'contextMenu',
    name: 'contextMenu',
    'data-testid': 'test-toolbar',
};
const mockToolbar = {
    id: 'test',
    name: 'test',
    'data-testid': 'mockToolbar',
    contextMenu: [mockContextMenu],
};
const mockData = [
    {
        id: 'open',
        name: 'Open',
    },
    {
        id: 'test',
        name: 'test',
        panel: [mockToolbar],
    },
];

// mock collapse
jest.mock('mo/components/collapse', () => {
    const originalModule = jest.requireActual('mo/components/collapse');
    return {
        ...originalModule,
        Collapse: jest
            .fn()
            // to mock originalModule once for ensure snapshot is the actual component
            .mockImplementationOnce(originalModule.Collapse)
            .mockImplementation(
                ({ data, onCollapseChange, onToolbarClick }) => (
                    <div data-testid="collapse">
                        {data.map((item, index) => {
                            const { panel = [] } = item;
                            return (
                                <span key={index}>
                                    <span
                                        data-testid={item.id}
                                        onClick={(e) =>
                                            onCollapseChange(item.id)
                                        }
                                    >
                                        collapse-{index}
                                    </span>
                                    {panel.map((p, index) => (
                                        <span
                                            data-testid={`${item.id}-${p.id}`}
                                            key={index}
                                            onClick={() =>
                                                onToolbarClick(p, item)
                                            }
                                        >
                                            panel-{index}
                                        </span>
                                    ))}
                                </span>
                            );
                        })}
                    </div>
                )
            ),
    };
});

describe('The Explorer Component', () => {
    afterEach(cleanup);

    test('Match Snapshot', () => {
        let component;
        act(() => {
            component = create(
                <Explorer data={mockData} headerToolBar={mockToolbar} />
            );
        });
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Should trigger onToolbarClick event', () => {
        expectFnCalled((mockFn) => {
            const { getByTestId } = render(
                <Explorer
                    data={mockData}
                    headerToolBar={mockToolbar}
                    onToolbarClick={mockFn}
                />
            );

            const toolbar = getByTestId(`${mockData[1].id}-${mockToolbar.id}`);
            fireEvent.click(toolbar);

            expect(mockFn.mock.calls[0][0]).toEqual(mockToolbar);
            expect(mockFn.mock.calls[0][1]).toEqual(mockData[1]);
        });
    });

    test('Should trigger onCollapseChange event', () => {
        expectFnCalled((mockFn) => {
            const { getByTestId } = render(
                <Explorer
                    data={mockData}
                    headerToolBar={mockToolbar}
                    onCollapseChange={mockFn}
                />
            );

            const target = mockData[1];
            const header = getByTestId(target.id);
            fireEvent.click(header);

            expect(mockFn.mock.calls[0][0]).toEqual(target.id);
        });
    });

    test('Should trigger onActionsContextMenuClick event', () => {
        expectFnCalled((mockFn) => {
            const { getByTestId } = render(
                <Explorer
                    data={mockData}
                    headerToolBar={mockToolbar}
                    onActionsContextMenuClick={mockFn}
                />
            );

            const target = getByTestId(mockToolbar['data-testid']);
            expect(target).toBeInTheDocument();
            fireEvent.contextMenu(target!);

            const toolbarItem = getByTestId(mockContextMenu['data-testid']);
            fireEvent.click(toolbarItem);

            expect(mockFn.mock.calls[0][1]).toEqual(
                expect.objectContaining(mockContextMenu)
            );
        });
    });

    test('Should trigger onClick event in header toolbar', () => {
        expectFnCalled((mockFn) => {
            const { getByTestId } = render(
                <Explorer
                    data={mockData}
                    headerToolBar={mockToolbar}
                    onClick={mockFn}
                />
            );

            const target = getByTestId(mockToolbar['data-testid']);

            expect(target).toBeInTheDocument();
            fireEvent.click(target!);

            const toolbarItem = getByTestId(mockContextMenu['data-testid']);
            fireEvent.click(toolbarItem);

            expect(mockFn.mock.calls[0][1]).toEqual(
                expect.objectContaining(mockToolbar)
            );
        });
    });
});
