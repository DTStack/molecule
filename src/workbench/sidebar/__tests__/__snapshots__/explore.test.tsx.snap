// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Explorer Component Match Snapshot 1`] = `
<div
  className="mo-sidebar-explorer"
>
  <header
    className="mo-sidebar__header"
  >
    <div
      className="mo-sidebar__title"
    >
      <h2>
        Explorer
      </h2>
    </div>
    <div
      className="mo-sidebar__toolbar"
    >
      <div
        className="mo-tool-bar"
      >
        <div
          className="mo-action-bar"
        >
          <ul
            className="mo-action-bar__container"
          >
            <li
              className="mo-action-bar__item"
              data-index={0}
              data-testid="mockToolbar"
              onClick={[Function]}
            >
              <span
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                test
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </header>
  <div
    className="mo-sidebar__content"
  >
    <div
      className="mo-collapse"
    >
      <div
        className="mo-split mo-split--horizontal"
      >
        <div
          className="mo-split__pane mo-collapse__pane"
          style={
            Object {
              "height": 26,
              "top": 0,
            }
          }
        >
          <div
            className="mo-collapse__item"
            data-collapse-id="open"
          >
            <div
              className="mo-collapse__header"
              onClick={[Function]}
              tabIndex={0}
            >
              <span
                className="codicon codicon-chevron-right"
              />
              <span
                className="mo-collapse__header__title"
              >
                Open
              </span>
              <div
                className="mo-collapse__extra"
              />
            </div>
            <div
              className="mo-collapse__content"
              data-collapse-index={0}
              tabIndex={0}
            />
          </div>
        </div>
        <div
          className="mo-split__sash mo-split__sash--disabled mo-split__sash--horizontal"
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="Resizer"
          style={
            Object {
              "height": 4,
              "top": 24,
            }
          }
        />
        <div
          className="mo-split__pane mo-collapse__pane"
          style={
            Object {
              "height": 26,
              "top": 26,
            }
          }
        >
          <div
            className="mo-collapse__item"
            data-collapse-id="test"
          >
            <div
              className="mo-collapse__header"
              onClick={[Function]}
              tabIndex={0}
            >
              <span
                className="codicon codicon-chevron-right"
              />
              <span
                className="mo-collapse__header__title"
              >
                test
              </span>
              <div
                className="mo-collapse__extra"
              />
            </div>
            <div
              className="mo-collapse__content"
              data-collapse-index={1}
              tabIndex={0}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
